microsoft365defender_connector.py

```diff
@@ -0,0 +1,1317 @@
+
+    # If code is not available
+    if not code:
+        return HttpResponse('Error while authenticating\n{0}'.format(json.dumps(request.GET)), content_type="text/plain", status=400)
+
+    state = _load_app_state(asset_id)
+
+    # If value of admin_consent is not available, value of code is available
+    try:
+        state[DEFENDER_CODE_STRING] = Microsoft365Defender_Connector().encrypt_state(code, asset_id=asset_id)
+        state[DEFENDER_STATE_IS_ENCRYPTED] = True
+    except Exception as e:
+        return HttpResponse("{}: {}".format(DEFENDER_DECRYPTION_ERROR, str(e)), content_type="text/plain", status=400)
+
+    _save_app_state(state, asset_id, None)
+
+    return HttpResponse('Code received. Please close this window, the action will continue to get new token.', content_type="text/plain")
+
+
+def _handle_rest_request(request, path_parts):
+    """ Handle requests for authorization.
+
+    :param request: Data given to REST endpoint
+    :param path_parts: parts of the URL passed
+    :return: dictionary containing response parameters
+    """
+
+    if len(path_parts) < 2:
+        return HttpResponse('error: True, message: Invalid REST endpoint request', content_type="text/plain", status=404)
+
+    call_type = path_parts[1]
+
+    # To handle authorize request in test connectivity action
+    if call_type == 'start_oauth':
+        return _handle_login_redirect(request, 'authorization_url')
+
+    # To handle response from microsoft login page
+    if call_type == 'result':
+        return_val = _handle_login_response(request)
+        asset_id = request.GET.get('state')
+        if asset_id and asset_id.isalnum():
+            app_dir = os.path.dirname(os.path.abspath(__file__))
+            auth_status_file_path = '{0}/{1}_{2}'.format(app_dir, asset_id, DEFENDER_TC_FILE)
+            real_auth_status_file_path = os.path.abspath(auth_status_file_path)
+            if not os.path.dirname(real_auth_status_file_path) == app_dir:
+                return HttpResponse("Error: Invalid asset_id", content_type="text/plain", status=400)
+            open(auth_status_file_path, 'w').close()
+            try:
+                uid = pwd.getpwnam('apache').pw_uid
+                gid = grp.getgrnam('phantom').gr_gid
+                os.chown(auth_status_file_path, uid, gid)
+                os.chmod(auth_status_file_path, '0664')
+            except Exception:
+                pass
+
+        return return_val
+    return HttpResponse('error: Invalid endpoint', content_type="text/plain", status=404)
+
+
+def _get_dir_name_from_app_name(app_name):
+    """ Get name of the directory for the app.
+
+    :param app_name: Name of the application for which directory name is required
+    :return: app_name: Name of the directory for the application
+    """
+
+    app_name = ''.join([x for x in app_name if x.isalnum()])
+    app_name = app_name.lower()
+    if not app_name:
+        app_name = 'app_for_phantom'
```

Comment:
Can you please explain why this is the default app_name?

Reply:
Hi @alexa-phantom,
As per my understanding the if condition at line 227 should never become true as the reason for this is _get_dir_name_from_app_name() method is getting called from _get_app_rest_url(). Here we are fetching the app_name variable from the app JSON (lines 620 - 623) and as long as the "name" key is present inside the app JSON this condition should never become true. But coincidentally this condition is present inside most Microsoft apps.

But if any reviewer knows the exact reason apart from this then please let me know.